File: LeetCode/33. 搜索旋转排序数组.ts
function search(nums: number[], target: number): number {
    let offset = getTurnPoint(nums);
    return binary_search(nums, target, offset);
};

function getTurnPoint(nums: number[]): number {
    let count = nums.length;
    let mod = nums.length;
    let cur = 0;
    while(count -- ){
        let pre = (cur - 1 + mod)%mod;
        let next = (cur + 1 + mod)%mod;
        if ((nums[next]-nums[cur])*(nums[cur]-nums[pre]) < 0) return cur+1;
        cur ++;
    }
    return 0;
}

function binary_search(nums:number[], target:number, offset:number):number {
    let low = 0, high = nums.length - 1;
    let mid = 0;
    let mod = nums.length;
    while(low <= mid) {
        mid = Math.floor(low + high / 2);
        let origin_pos = (mid - offset + mod)%nums.length;
        if (nums[origin_pos] == target) return mid;
        else if(nums[origin_pos] < target) low = mid + 1;
        else high = mid - 1;
    }
    return -1;
}

File: Week_04/33. 搜索旋转排序数组.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        arrSize = len(nums)
        # 计算出数组起点的偏移量offset,也就是所有元素的偏移量offset
        offset = self.getTurnPoint(nums)
        low = 0
        high = arrSize-1
        while low <= high:
            mid = int((low + high) / 2)
            # 数组进过翻转后已经产生了偏移，实际位置是(mid+offset)%arrSize
            if nums[(mid+offset)%arrSize] == target: return (mid+offset)%arrSize
            elif nums[(mid+offset)%arrSize] < target: low = mid + 1
            else: high = mid - 1
        return -1

    def getTurnPoint(self, nums: List[int]) -> int:
        arrSize = len(nums)
        for i in range(arrSize):
            if nums[i] <= nums[(i-1)%arrSize]: 
                return i
    